src/analyze/analyze-unit-shell.c

```
@@ -0,0 +1,109 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to read PID from reply: %m");
+
+        _cleanup_close_ int mntns_fd = -EBADF, root_fd = -EBADF, pidns_fd = -EBADF, netns_fd = -EBADF, userns_fd = -EBADF;
+        r = namespace_open(
+                        pid,
+                        &pidns_fd,
+                        &mntns_fd,
+                        &netns_fd,
+                        &userns_fd,
+                        &root_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to retrieve FDs of the target process' namespace: %m");
+
+        _cleanup_strv_free_ char **args = argc > 2 ? strv_copy(strv_skip(argv, 2)) : NULL;
+        pid_t child;
+        r = namespace_fork(
+                        "(unit-shell-ns)",
+                        "(unit-shell)",
+                        NULL, /* except_fds[] */
+                        0, /* n_except_fds */
+                        FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL,
+                        pidns_fd,
+                        mntns_fd,
+                        netns_fd,
+                        userns_fd,
+                        root_fd,
+                        &child);
+        if (r < 0)
+                return log_error_errno(r, "Failed to fork the namespace: %m");
+
+        if (r == 0) {
+                if (args) {
+                        r = execvp(args[0], args);
+                        if (r < 0)
+                                return log_error_errno(r, "execvp(%s, [%s...]) failed: %m", args[0], *args);
+                }
+                else {
+                        execl(DEFAULT_USER_SHELL, "-" DEFAULT_USER_SHELL_NAME, NULL);
+                        if (!streq(DEFAULT_USER_SHELL, "/bin/bash")) {
+                                r = execl("/bin/bash", "-bash", NULL);
+                                if (r < 0)
+                                        return log_error_errno(r, "execl(\"/bin/bash\", \"-bash\") failed: %m");
```

- Please do not return.
- `r` -> `errno`
- the failure will be ignored, and the fallback shell will be used in the below, hence downgrade the log level to debug, and append ', ignored' in the log message.